src/main/ank/2.3-free-monoids.md

```diff
@@ -0,0 +1,90 @@
+```Haskell
+class Monoid m where
+  mempty :: m
+  mappend :: m -> m -> m
+```
+```kotlin:ank:silent
+interface Monoid<A> {
+  fun empty(): A
+  fun combine(a: A, b: A): A
+}
+```
+................
+```Haskell
+instance Monoid [a] where
+  mempty = []
+  mappend = (++)
+```
+```kotlin
+import arrow.extension
+import arrow.typeclasses.MonoidOf
+
+@extension interface Monoid<A>: MonoidOf<List<A>> {
+    fun empty() = emptyList<A>()
+    fun List<A>.combine(other: List<A>) = this + other
+}
+```
+................
+```Haskell
+2 * 3 = 6
+[2] ++ [3] = [2, 3] // not the same as [6]
+```
+```kotlin
+2 * 3 == 6
+listOf(2) + listOf(3) == listOf(2, 3) // not the same as listOf(6)
+```
+
+................
+```Haskell
+h (a * b) = h a * h b
+```
+```kotlin
+h(a * b) == h(a) * h(b)
+```
+
+................
+```Haskell
+[2] ++ [3] = [2, 3]
+```
+```kotlin
+listOf(2) + listOf(3) == listOf(2, 3)
+```
+
+................
+```Haskell
+2 * 3 = 6
+```
+```kotlin
+2 * 3 == 6
+```
+................
+```Haskell
+p :: x -> U m
+```
+```kotlin
+val p : (X) -> Kind<U, M>
```

Comment:
```
from the book:
We start with a set of generators, x. That's a set in Set. 
The pattern we are going to match consists of monoid m - an object of Mon - and a function p in Set,
where U is our forgetful functor from Mon to Set. 
This is a weird heterogeneous pattern - half in Mon and half in Set.
The idea is that the function p will identitfy the set of generators inside the X-ray image of m. 
(Page 212)
```
@nomisRev: I am not sure, but it looks like this should be something like this:
```suggestion
val p : (A) -> U<M>
```

Reply:
This is actually not valid Kotlin code.

```
fun <X, U, M> p(a: A): Kind<U, M>
```

There is no way to use generics unless they are defined in a function, or the lambda lives inside a class or interface that defines those generics.